File: collect_app/src/main/java/org/odk/collect/android/listeners/SwipeHandler.java
package org.odk.collect.android.listeners;

import android.content.Context;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.widget.NestedScrollView;

import org.odk.collect.android.utilities.FlingRegister;
import org.odk.collect.androidshared.utils.ScreenUtils;
import org.odk.collect.settings.keys.ProjectKeys;
import org.odk.collect.shared.settings.Settings;

public class SwipeHandler {

    private final GestureDetector gestureDetector;
    private final OnSwipeListener onSwipe;
    private View view;
    private boolean allowSwiping = true;
    private boolean beenSwiped;
    private final Settings generalSettings;

    public interface OnSwipeListener {
        void onSwipeBackward();
        void onSwipeForward();
    }

    public SwipeHandler(Context context, Settings generalSettings) {
        gestureDetector = new GestureDetector(context, new GestureListener());
        this.onSwipe = (OnSwipeListener) context;
        this.generalSettings = generalSettings;
    }

    public void setView(View view) {
        this.view = view;
    }

    public void setAllowSwiping(boolean allowSwiping) {
        this.allowSwiping = allowSwiping;
    }

    public void setBeenSwiped(boolean beenSwiped) {
        this.beenSwiped = beenSwiped;
    }

    public boolean beenSwiped() {
        return beenSwiped;
    }

    public GestureDetector getGestureDetector() {
        return gestureDetector;
    }

    public class GestureListener implements GestureDetector.OnGestureListener {

        @Override
        public boolean onDown(MotionEvent event) {
            return false;
        }

        @Override
        public void onShowPress(MotionEvent e) {

        }

        @Override
        public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
            // The onFling() captures the 'up' event so our view thinks it gets long pressed. We don't want that, so cancel it.
            if (view != null) {
                view.cancelLongPress();
            }
            return false;
        }

        @Override
        public void onLongPress(MotionEvent e) {

        }

        @Override
        public boolean onSingleTapUp(MotionEvent e) {
            return false;
        }

        @Override
        public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
            if (view == null) {
                return false;
            }

            FlingRegister.flingDetected();

            if (e1 != null && e2 != null
                    && generalSettings.getString(ProjectKeys.KEY_NAVIGATION).contains(ProjectKeys.NAVIGATION_SWIPE)
                    && allowSwiping) {
                // Looks for user swipes. If the user has swiped, move to the appropriate screen.

                // For all screens a swipe is left/right of at least .25" and up/down of less than .25" OR left/right of > .5"
                int xpixellimit = (int) (ScreenUtils.xdpi(view.getContext()) * .25);
                int ypixellimit = (int) (ScreenUtils.ydpi(view.getContext()) * .25);

                if (view != null && view.shouldSuppressFlingGesture()) {
                    return false;
                }

                if (beenSwiped) {
                    return false;
                }

                float diffX = Math.abs(e1.getX() - e2.getX());
                float diffY = Math.abs(e1.getY() - e2.getY());

                if (view != null && canScrollVertically() && getGestureAngle(diffX, diffY) > 30) {
                    return false;
                }

                if ((diffX > xpixellimit && diffY < ypixellimit) || diffX > xpixellimit * 2) {
                    beenSwiped = true;
                    if (e1.getX() > e2.getX()) {
                        onSwipe.onSwipeForward();
                    } else {
                        onSwipe.onSwipeBackward();
                    }
                    return true;
                }
            }

            return false;
        }

        private double getGestureAngle(float diffX, float diffY) {
            return Math.toDegrees(Math.atan2(diffY, diffX));
        }

        public boolean canScrollVertically() {
            NestedScrollView scrollView = view.getVerticalScrollView();

            if (scrollView != null) {
                int screenHeight = scrollView.getHeight();
                int viewHeight = scrollView.getChildAt(0).getHeight();
                return viewHeight > screenHeight;
            } else {
                return false;
            }
        }
    }

    public abstract static class View extends FrameLayout {
        public View(@NonNull Context context) {
            super(context);
        }

        public abstract boolean shouldSuppressFlingGesture();

        @Nullable
        public abstract NestedScrollView getVerticalScrollView();
    }
}

File: collect_app/src/main/java/org/odk/collect/android/listeners/SwipeHandler.kt
package org.odk.collect.android.listeners

import android.content.Context
import android.view.GestureDetector
import android.view.MotionEvent
import android.widget.FrameLayout
import androidx.core.widget.NestedScrollView
import org.odk.collect.android.utilities.FlingRegister
import org.odk.collect.androidshared.utils.ScreenUtils
import org.odk.collect.settings.keys.ProjectKeys
import org.odk.collect.shared.settings.Settings
import kotlin.math.abs
import kotlin.math.atan2

class SwipeHandler(context: Context, generalSettings: Settings) {
    val gestureDetector: GestureDetector
    private val onSwipe: OnSwipeListener
    private var view: View? = null
    private var allowSwiping = true
    private var beenSwiped = false
    private val generalSettings: Settings

    interface OnSwipeListener {
        fun onSwipeBackward()
        fun onSwipeForward()
    }

    init {
        gestureDetector = GestureDetector(context, GestureListener())
        onSwipe = context as OnSwipeListener
        this.generalSettings = generalSettings
    }

    fun setView(view: View?) {
        this.view = view
    }

    fun setAllowSwiping(allowSwiping: Boolean) {
        this.allowSwiping = allowSwiping
    }

    fun setBeenSwiped(beenSwiped: Boolean) {
        this.beenSwiped = beenSwiped
    }

    fun beenSwiped() = beenSwiped

    inner class GestureListener : GestureDetector.OnGestureListener {
        override fun onDown(event: MotionEvent) = false
        override fun onSingleTapUp(e: MotionEvent) = false

        override fun onShowPress(e: MotionEvent) = Unit
        override fun onLongPress(e: MotionEvent) = Unit

        override fun onScroll(e1: MotionEvent, e2: MotionEvent, distanceX: Float, distanceY: Float): Boolean {
            // The onFling() captures the 'up' event so our view thinks it gets long pressed. We don't want that, so cancel it.
            view?.cancelLongPress()
            return false
        }

        override fun onFling(e1: MotionEvent, e2: MotionEvent, velocityX: Float, velocityY: Float): Boolean {
            if (view == null) {
                return false
            }

            FlingRegister.flingDetected()

            if (generalSettings.getString(ProjectKeys.KEY_NAVIGATION)!!.contains(ProjectKeys.NAVIGATION_SWIPE) && allowSwiping) {
                // Looks for user swipes. If the user has swiped, move to the appropriate screen.

                // For all screens a swipe is left/right of at least .25" and up/down of less than .25" OR left/right of > .5"
                val xpixellimit = (ScreenUtils.xdpi(view!!.context) * .25).toInt()
                val ypixellimit = (ScreenUtils.ydpi(view!!.context) * .25).toInt()

                if (view != null && view!!.shouldSuppressFlingGesture()) {
                    return false
                }

                if (beenSwiped) {
                    return false
                }

                val diffX = abs(e1.x - e2.x)
                val diffY = abs(e1.y - e2.y)

                if (view != null && canScrollVertically() && getGestureAngle(diffX, diffY) > 30) {
                    return false
                }

                if (diffX > xpixellimit && diffY < ypixellimit || diffX > xpixellimit * 2) {
                    beenSwiped = true
                    if (e1.x > e2.x) {
                        onSwipe.onSwipeForward()
                    } else {
                        onSwipe.onSwipeBackward()
                    }
                    return true
                }
            }
            return false
        }

        private fun getGestureAngle(diffX: Float, diffY: Float): Double {
            return Math.toDegrees(atan2(diffY.toDouble(), diffX.toDouble()))
        }

        private fun canScrollVertically(): Boolean {
            val scrollView = view!!.verticalScrollView

            return if (scrollView != null) {
                val screenHeight = scrollView.height
                val viewHeight = scrollView.getChildAt(0).height
                viewHeight > screenHeight
            } else {
                false
            }
        }
    }

    abstract class View(context: Context) : FrameLayout(context) {
        abstract fun shouldSuppressFlingGesture(): Boolean
        abstract val verticalScrollView: NestedScrollView?
    }
}
